---
title: Morphofunctional changes at the active zone during synaptic vesicle exocytosis
keywords:
- synapse
- cryo-electron tomography
- synaptic vesicles
- deep learning
- segmentation
- post-processing
lang: en-US
date-meta: '2021-11-29'
author-meta:
- Amin Khosrozadeh
- Raphaela Seeger
- Julika Radecke
- Guillaume Witz
- Vladan Lučić
- Jakob B. Sørensen
- Benoît Zuber
header-includes: |-
  <!--
  Manubot generated metadata rendered from header-includes-template.html.
  Suggest improvements at https://github.com/manubot/manubot/blob/main/manubot/process/header-includes-template.html
  -->
  <meta name="dc.format" content="text/html" />
  <meta name="dc.title" content="Morphofunctional changes at the active zone during synaptic vesicle exocytosis" />
  <meta name="citation_title" content="Morphofunctional changes at the active zone during synaptic vesicle exocytosis" />
  <meta property="og:title" content="Morphofunctional changes at the active zone during synaptic vesicle exocytosis" />
  <meta property="twitter:title" content="Morphofunctional changes at the active zone during synaptic vesicle exocytosis" />
  <meta name="dc.date" content="2021-11-29" />
  <meta name="citation_publication_date" content="2021-11-29" />
  <meta name="dc.language" content="en-US" />
  <meta name="citation_language" content="en-US" />
  <meta name="dc.relation.ispartof" content="Manubot" />
  <meta name="dc.publisher" content="Manubot" />
  <meta name="citation_journal_title" content="Manubot" />
  <meta name="citation_technical_report_institution" content="Manubot" />
  <meta name="citation_author" content="Amin Khosrozadeh" />
  <meta name="citation_author_institution" content="Institute of Anatomy, University of Bern, Bern, Switzerland" />
  <meta name="citation_author_institution" content="Graduate School for Cellular and Biomedical Sciences, University of Bern" />
  <meta name="citation_author_orcid" content="XXXX-XXXX-XXXX-XXXX" />
  <meta name="citation_author" content="Raphaela Seeger" />
  <meta name="citation_author_institution" content="Institute of Anatomy, University of Bern, Bern, Switzerland" />
  <meta name="citation_author_institution" content="Graduate School for Cellular and Biomedical Sciences, University of Bern" />
  <meta name="citation_author_orcid" content="XXXX-XXXX-XXXX-XXXX" />
  <meta name="citation_author" content="Julika Radecke" />
  <meta name="citation_author_institution" content="Institute of Anatomy, University of Bern, Bern, Switzerland" />
  <meta name="citation_author_institution" content="Department of Neuroscience, Faculty of Health and Medical Science , 2200 Copenhagen N, University of Copenhagen, Copenhagen, Denmark" />
  <meta name="citation_author_institution" content="Diamond Light Source Ltd, Didcot, Oxfordshire, United Kingdom" />
  <meta name="citation_author_orcid" content="0000-0002-5815-5537" />
  <meta name="citation_author" content="Guillaume Witz" />
  <meta name="citation_author_institution" content="Science IT Service, University of Bern, Bern, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0003-1562-4265" />
  <meta name="citation_author" content="Vladan Lučić" />
  <meta name="citation_author_institution" content="Max-Planck-Institute of Biochemistry, Am Klopferspitz 18, 82152 Martinsried, Germany" />
  <meta name="citation_author_orcid" content="0000-0003-3698-7436" />
  <meta name="citation_author" content="Jakob B. Sørensen" />
  <meta name="citation_author_institution" content="Department of Neuroscience, University of Copenhagen, Blegdamsvej 3B, 2200 Copenhagen N, Denmark" />
  <meta name="citation_author_orcid" content="0000-0001-5465-3769" />
  <meta name="citation_author" content="Benoît Zuber" />
  <meta name="citation_author_institution" content="Institute of Anatomy, University of Bern, Bern, Switzerland" />
  <meta name="citation_author_orcid" content="0000-0001-7725-5579" />
  <link rel="canonical" href="https://elatella.github.io/deep-prepyto-paper/" />
  <meta property="og:url" content="https://elatella.github.io/deep-prepyto-paper/" />
  <meta property="twitter:url" content="https://elatella.github.io/deep-prepyto-paper/" />
  <meta name="citation_fulltext_html_url" content="https://elatella.github.io/deep-prepyto-paper/" />
  <meta name="citation_pdf_url" content="https://elatella.github.io/deep-prepyto-paper/manuscript.pdf" />
  <link rel="alternate" type="application/pdf" href="https://elatella.github.io/deep-prepyto-paper/manuscript.pdf" />
  <link rel="alternate" type="text/html" href="https://elatella.github.io/deep-prepyto-paper/v/760cbe767fea7aff665cafca1c24f22ccad8b09d/" />
  <meta name="manubot_html_url_versioned" content="https://elatella.github.io/deep-prepyto-paper/v/760cbe767fea7aff665cafca1c24f22ccad8b09d/" />
  <meta name="manubot_pdf_url_versioned" content="https://elatella.github.io/deep-prepyto-paper/v/760cbe767fea7aff665cafca1c24f22ccad8b09d/manuscript.pdf" />
  <meta property="og:type" content="article" />
  <meta property="twitter:card" content="summary_large_image" />
  <link rel="icon" type="image/png" sizes="192x192" href="https://manubot.org/favicon-192x192.png" />
  <link rel="mask-icon" href="https://manubot.org/safari-pinned-tab.svg" color="#ad1457" />
  <meta name="theme-color" content="#ad1457" />
  <!-- end Manubot generated metadata -->
bibliography:
- content/manual-references.json
manubot-output-bibliography: output/references.json
manubot-output-citekeys: output/citations.tsv
manubot-requests-cache-path: ci/cache/requests-cache
manubot-clear-requests-cache: false
...






<small><em>
This manuscript
([permalink](https://elatella.github.io/deep-prepyto-paper/v/760cbe767fea7aff665cafca1c24f22ccad8b09d/))
was automatically generated
from [elatella/deep-prepyto-paper@760cbe7](https://github.com/elatella/deep-prepyto-paper/tree/760cbe767fea7aff665cafca1c24f22ccad8b09d)
on November 29, 2021.
</em></small>

## Authors



+ **Amin Khosrozadeh**^[\*](#equal_contribution)^<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [XXXX-XXXX-XXXX-XXXX](https://orcid.org/XXXX-XXXX-XXXX-XXXX)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [ameen-khosrowzadeh](https://github.com/ameen-khosrowzadeh)<br>
  <small>
     Institute of Anatomy, University of Bern, Bern, Switzerland; Graduate School for Cellular and Biomedical Sciences, University of Bern
     · Funded by Grant XXXXXXXX
  </small>

+ **Raphaela Seeger**^[\*](#equal_contribution)^<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [XXXX-XXXX-XXXX-XXXX](https://orcid.org/XXXX-XXXX-XXXX-XXXX)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [elatella](https://github.com/elatella)<br>
  <small>
     Institute of Anatomy, University of Bern, Bern, Switzerland; Graduate School for Cellular and Biomedical Sciences, University of Bern
  </small>

+ **Julika Radecke**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0002-5815-5537](https://orcid.org/0000-0002-5815-5537)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [julikaradecke](https://github.com/julikaradecke)<br>
  <small>
     Institute of Anatomy, University of Bern, Bern, Switzerland; Department of Neuroscience, Faculty of Health and Medical Science , 2200 Copenhagen N, University of Copenhagen, Copenhagen, Denmark; Diamond Light Source Ltd, Didcot, Oxfordshire, United Kingdom
  </small>

+ **Guillaume Witz**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-1562-4265](https://orcid.org/0000-0003-1562-4265)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [guiwitz](https://github.com/guiwitz)<br>
  <small>
     Science IT Service, University of Bern, Bern, Switzerland
  </small>

+ **Vladan Lučić**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0003-3698-7436](https://orcid.org/0000-0003-3698-7436)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [vladanl](https://github.com/vladanl)<br>
  <small>
     Max-Planck-Institute of Biochemistry, Am Klopferspitz 18, 82152 Martinsried, Germany
  </small>

+ **Jakob B. Sørensen**<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-5465-3769](https://orcid.org/0000-0001-5465-3769)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [JBSorensen](https://github.com/JBSorensen)<br>
  <small>
     Department of Neuroscience, University of Copenhagen, Blegdamsvej 3B, 2200 Copenhagen N, Denmark
     · Funded by Novo Nordisk Fonden, NNF17OC0028516.; Carlsbergfondet, CF17-0875; Independent Research Fond Denmark, 8020-00228A; Lundbeckfonden, R277-2018-802
  </small>

+ **Benoît Zuber**^[✉](#correspondence)^<br>
    ![ORCID icon](images/orcid.svg){.inline_icon}
    [0000-0001-7725-5579](https://orcid.org/0000-0001-7725-5579)
    · ![GitHub icon](images/github.svg){.inline_icon}
    [aseedb](https://github.com/aseedb)<br>
  <small>
     Institute of Anatomy, University of Bern, Bern, Switzerland
     · Funded by Swiss National Science Foundation, 179520; ERA-NET NEURON, NEURON-119
  </small>


::: {#correspondence}
✉ Address correspondence to <benoit.zuber@ana.unibe.ch> and <jakobbs@sund.ku.dk>.
:::
::: {#equal_contribution}
\* These authors contributed equally.
:::


## Abstract {.page_break_before}

The communication between neurons is based on the exocytosis of neurotransmitter filled vesicles. Even small disbalances in this delicate system can result in different disorders. Numerous questions are still unanswered about synaptic vesicle exocytosis regulation as well as about which proteins affect this fundamental process. Hundreds to thousands of vesicles are present at the synapse and most of them are interconnected. There is a number of evidence suggesting that the connectivity between the vesicles influences synaptic vesicle exocytotic activity and that connectivity can be regulated. To address these issues, cryo-electron tomography is a method of choice. Yet inevitable quantitative studies require accurate segmentation of both synaptic vesicles and the interconnecting filaments.

This segmentation process has been widely recognized as a bottleneck by the community. When working with synapses, an incredibly tedious cellular feature to segment are synaptic vesicles. Even though they have a very homogeneous shape and size, which makes it easy to distinguish them from other cellular features, their high occurrence makes their segmentation challenging (one average-sized synapse or synaptosomes can easily contain between 300-500 synaptic vesicles).

Several attempts at utilizing machine learning algorithms  to automize the segmentation of synaptic vesicles yielded in unfavorable results. We analyzed these approaches and realized that most algorithms solely worked with the 2D image planes within the tomogram, but didn’t tap into their 3D information and therefore increased the weight of the missing wedge.

We designed a framework that consists of two parts: 1) U-Net convolutional network to obtain mask for SVs following by 2) post-processing step which fine-tunes the mask and provides entire 3D-vesicles. A deep-learning model has been trained with manually segmented synaptic vesicles, which have been entirely annotated as 3D objects. By post-processing, we achieved to transfer the vesicle recognition through this pipeline. The evaluation metrics like the DICE value suggest a very high success rate.


## Introduction

The rise of machine learning is sweeping the technological landscape.
Its use cases are highly diverse and are just starting to reach the scientific community.
Especially the graphical deep-learning approaches have been quite popular [@doi:10.1056/NEJMra1814259].
But also in the field of cryo electron microscopy has recently recognized the benefits of machine/deep-learning approaches [@doi:10.1093/jmicro/dfz036].

The method of cryo-electron tomography (ET) utilises a series of tilted two-dimensional (2D) images to reconstruct a three-dimensional (3D) image of the imaged sample [@doi:10.1042/BCJ20200715].
It brings many advantages compared to other cryo-electron microscopy (EM) methods, as it yields images at in situ conditions.
Therefore cryo ET can help to reveal the natural cellular environment of protein(-complexes), their function as well as their structure.
In order to analyze the cellular context within a tomogram, cellular features are being segmented. Popular programs for manual segementation are IMOD and Amira.


## Results


### Comparison of manual segmentation with automatic deep-learning based segmentation



## Discussion






## Materials and methods
### Constructs and viruses
SNAP-25B was N-terminally fused to GFP and cloned into a pLenti construct with a CMV promoter [@doi:10.1523/JNEUROSCI.5092-06.2007].
Mutations were made using the QuikChange II XL kit (Agilent).
The mutations were verified by sequencing and have been published before [@doi:10.1016/j.celrep.2019.01.103].
The preparation of lentiviral particles followed standard protocols.

### Animals
Synaptosomes were prepared from adult male or female Wistar rat obtained from the central animal facilities of the Department of Biomedical Research of the University of Bern.
Adult male or female Wistar rats at an age of 6-8 weeks were slightly stunned by CO2 and quickly decapitated with a guillotine.
The procedures used were in accordance with the Swiss Veterinary Law guidelines.
Heterozygous SNAP-25 KO C57/Bl6-mice were routinely backcrossed to Bl6 to generate new heterozygotes.
The strain was kept in the heterozygous condition and timed heterozygous crosses and caesarean section were used to recover knockout embryos at embryonic day 18 (E18). 
Pregnant females were killed by cervical dislocation; embryos of either sex were collected and killed by decapitation. Permission to keep and breed SNAP-25 mice was obtained from the Danish Animal Experiments Inspectorate and followed institutional guidelines as overseen by the Institutional Animal Care and Use Committee (IACUC).
CD1 outbred mice were used to create astrocytic cultures.
Newborns (P0-P2) of either sex were used.
Pups were killed by decapitation.

### Synaptosome preparation
Rat synaptosomes were prepared as previously described [@doi:10.1038/nprot.2008.171], with some modifications.
The cerebral cortex and the hippocampi were removed in sucrose buffer (SEH: 0.32 M sucrose, 1 mM EDTA, 10 mM HEPES; HEPES, #H4034, Sigma-Aldrich Corporate Offices. St. Louis, MO, USA) on ice. 
 Homogenization of the tissue was done in SEH with a Potter-Elvehjem grinder (#358011, Wheaton. Millville, New Jersey, USA), four strokes at the bottom and 6 from top to bottom were applied to the tissue at a speed of 800 turns/min as described in [@doi:10.1038/nprot.2008.171]. 
 The whole process from decapitation to homogenization was done within 2-3 min, to obtain functional synaptosomes. 
 Homogenized tissue was then centrifuged at 1000 g for 10 min at 4°C to remove meninges and blood vessels. 
 The resulting supernatant containing synaptosomes, but also gliosomes and mitochondria was then added to a discontinuous, isoosmotic Percoll (#P1644, Sigma) gradient with 5%, 10% and 23% in 0.32 M sucrose, 1 mM EDTA in centrifuge tubes (#344060, Beckman Coulter). 
 The samples were spun in an ultracentrifuge (rotor: SW 40 Ti; Beckman Coulter. Nyon, Switzerland) at 16400 rpm for 12 min at 4°C. 
 The layer with the highest amount of functional synaptosomes was between 10-23 % [@doi:10.1038/nprot.2008.171]. 
 The layer was carefully taken out and diluted 1:10 in HEPES buffered medium (HBM; 140 mM NaCl, 5 mM KCl, 5 mM NaHCO~3~, 1.2 mM Na~2~HPO~4~, 1 mM MgCl~2~, 10 mM Glucose, 20 mM HEPES). 
 The obtained solution was further spun with an ultracentrifuge (rotor 45 Ti; Beckman Coulter) at 11200 rpm for 20 min at 4°C. 
 The pellet was carefully and quickly aspirated with a Pasteur pipette to avoid mixture with the solution and then diluted in HBM.

### Preparation of astrocytic and neuronal culture
The procedure has been published before [@doi:10.3791/50783].
Glial cells were ready to be used after 10 days. once they were triturated and counted with a Buerker chamber, 100,000 cells/ml were plated onto untreated 12 well plates containing 10% DMEM.
Astrocytes were isolated from CD1 outbred mice (P0-P2).
Pups were killed by decapitation and heads were placed in HBSS-HEPES medium (HBSS supplemented with 1 M HEPES).
The cortices were isolated from the brains and the meninges were removed (dura, pia and arachnoid mater).
The cortices were chopped into smaller fragments and transferred to a tube containing 0.25% trypsin dissolved in Dulbecco’s MEM (DMEM)supplemented with 10% foetal calf serum, 20000 IU penicillin, 20 mg streptomycin and 1% MEM non-essential amino acids. Fragments were incubated for 15 min at 37°C.
Subsequently, inactivation medium (12.5 mg albumin + 12.5 mg trypsin-inhibitor in 10% DMEM) was added and the tissue washed with HBSS-HEPES. 
Tissue was triturated until a smooth cloudy suspension appeared. 
Cells were plated in 80 cm^2^ flasks with pre-warmed DMEM, one hemisphere per flask, and stored at 37°C with 5% CO~2~. Glial cells were ready to be used after 10 days.
Glial cells were washed with pre-warmed HBSS-HEPES.
Trypsin was added and the flasks were incubated at 37°C for 10 min.
Cells were triturated and counted with a Buerker chamber before plating 100,000 cells/ml on untreated 12-well plates containing DMEM.
After 2 days, neurons were plated. 

Hippocampal neurons were isolated from either E18 SNAP-25 KO.
The SNAP-25 KO pups were obtained by pairing two heterozygote animals, and the embryos were recovered at E18 by caesarean section. 
Pups were selected based on the absence of motion after tactile stimulation and bloated neck [@doi:10.1038/nn783]; the genotype was confirmed by PCR in all cases. 
The pups were killed by decapitation and heads were put in HBSS-HEPES medium. The cortices were isolated from the brains and the meninges were removed. 
The hippocampi were cut from the cortices before being transferred to a tube containing 0.25% trypsin dissolved in HBSS-HEPES solution.
Fragments were incubated for 20 min at 37°C. 
Afterwards, the tissue was washed with HBSS-HEPES. 
The hippocampi were triturated and the cell count was determined with a Buerker chamber. 
20 µl of solution containing 250,000 cells/ml were plated onto the flame sterilized gold R2/2 or R2/1 EM grids as previously described in [@doi:10.3791/50783]. 
Following a 30 min incubation at 37°C, the grid was transferred into the 12 well plate containing the astrocytes and medium was replaced with NB medium (Neurobasal with 2% B-27, 1 M HEPES, 0.26% lutamax, 14.3 mM β-mercaptoethanol, 20000 IU penicillin, 20 mg streptomycin) for the E18 pups or NB-A medium (Neurobasal-A with 2% B-27, 1% Glutamax , 20000 IU Penicillin, 20 mg Streptomycin) for the P0-P1.
Between 4 h and 1 day later, lentiviral particles carrying either SNAP-25-WT, SNAP-25-4E, or SNAP-25-4K constructs were added to the culture.
The cultures were incubated for 12 to 14 days before being plunge frozen.

### Plunge freezing and spray-mixing
Rat synaptosomes were prepared for plunge freezing and spray-mixing as follows.
The following steps from incubation to plunge freezing were all done at room temperature (RT), equivalent to 23-25°C. 
 The synaptosomal solution was incubated with calcein blue AM (#C1429, Molecular Probes-Thermo Fisher Scientific. Waltham, MA, USA) 30 min prior to plunge freezing to visualize the cytosol of functional – esterase containing – cellular compartments such as synaptosomes. 
 Additionally, 1.3 mM CaCl2 and 10 nm gold fiducials were added (gold fiducials, #s10110/8. AURION Immuno Gold Reagents & Accessories. Wageningen, The Netherlands). 
CaCl2 is necessary to trigger exocytosis and gold fiducials are important to align the acquired tilt series for tomogram reconstruction. 
The sprayed solution contained 1 mM CaCl2 and 52 mM KCl in HBM to depolarize synaptosomes and trigger exocytosis. 
It also contained fluorescein (#46955, Sigma) to trace the spray droplets on the EM grid in cryo-FM. 
The synaptosomal solution was applied to a 200-mesh lacey finder carbon film grid (#AGS166-H2. Agar Scientific. Elektron Technology UK Ltd. Stansted, UK). 
Excess liquid on the grid was removed by blotting with a filter paper and the grid was immediately plunge frozen in liquid ethane with a homebuilt plunge freezer and was sprayed on the fly. 
The plunge freezer and the spraying device (atomizer) were computer controlled with a LabView script (National Instruments Corporation. Mopac Expwy Austin, TX, USA). 
The spraying device was set similarly to the device in [@doi:10/ctwp5j]. 
Nitrogen gas pressure necessary to drive spraying was set to 2.5 bar. 
The grid was set to pass in front of the spray nozzle at a distance of 3-4 mm. 
The plunge freezer was accelerated to 0.75 m/s and the minimum spray delay was ~7 ms (Table S3).
The atomizer sprays scattered droplets of various size on the EM grid.
During the time lapse between spraying and freezing the content of the droplets spreads by diffusion.
KCl diffuses approximately 4x faster than fluorescein.
Cryo-ET imaging was done within the diffusion distance of KCl but outside of the visible spray droplet because the center of the spray droplet would usually be too thick for imaging.
This reduces the effective stimulation duration to anything between 0 ms and less than the given spray-freeze delay.
Moreover, through diffusion, KCl concentration rapidly rises and then decreases.
Hence synaptosomes are not permanently depolarized.

After 12 to 14 days of incubation grids with mouse neurons were plunge frozen with a Vitrobot (Thermofisher Scientific, Mark IV) with a blot time of 3 s and a blot force of -10. 
Wait time and drain time were not used. 
Humidity was set to 100% at 4°C. 
4 µl undiluted 10 nm BSA gold tracer (Aurion) was added directly onto the grid prior to plunge freezing.

### Cryo-fluorescence microscopy
After plunge freezing, rat synaptsome samples were imaged at the fluorescent microscope under cryo conditions, with a Zeiss Axio Scope.A1, equipped with an AxioCam MRm camera (Carl Zeiss AG, Germany), and a fluorescence lamp (HXP 120 C).
The correlative microscopy stage (#CMS196, Linkam Scientific Instruments, UK) was cooled down to -190°C by liquid nitrogen and the frozen EM grid was placed into the chamber of the cryostage on a bridge that was not submerged in liquid nitrogen and was close to the objective, where the temperature was around -150°C.
The filter set used for imaging fluorescein was #38 (#000000-1031-346, Zeiss) (BP 470/40, FT 495, BP 525/50; corresponds to GFP) and the one for calcein blue AM was #49 (#488049-9901-000, Zeiss) (G 365, FT 395, BP 445/50; corresponds to DAPI).
The objective used was either a 10x (#420941-9911, NA = 0.25 Ph1, Zeiss) or a 50x (#422472-9900, NA = 0.55 Dic, Zeiss), the acquisition software used was AxioVision (AxioVs40x64 V 4.8.3.0, Zeiss) and the processing software was ZEN lite (Zeiss).

### Cryo-electron microscopy
Following cryo-FM, the rat synaptosome grids were mounted in a cryo-holder (Gatan, Pleasonton, CA, USA) and transferred to a Tecnai F20 (FEI, Eindhoven, The Netherlands) which was set to low dose conditions, operated at 200 kV, and equipped with a field emission gun.
Images were recorded with a 2k x 2k CCD camera (Gatan) mounted after a GIF Tridiem post-column filter (Gatan) operated in zero-loss mode.
The sample was kept at about -180°C.
Tilt series were acquired using SerialEM [@doi:10.1016/j.jsb.2005.07.007] for automated acquisition recorded typically from -50° to 50° with a 2° angular increment and an unbinned pixel size of 0.75 or 1.2 nm.
Due to sample thickness (400-700 nm), tomograms were usually not recorded with higher tilt angles.
Defocus was set between -8 to -12 µm and the total electron dose used was about 80-100 e^-^/Å^2^.
Some tomograms were acquired at a Titan Krios equipped with a K2 direct electron detector (Gatan) without energy filter.
The K2 camera was operated in superresolution counting mode and between 8-40 frames per tilt angle were taken.
Tilt series were acquired using the Latitude software (Gatan) for automated acquisition recorded typically from -60° to 60° with a 2° angular increment and an unbinned pixel size of 0.6 nm.
Defocus was set between -8 to -12 µm and the total electron dose used was about 80-100 e^-^/Å^2^.
Prior to image processing the frames at each tilt angle, frames were aligned and averaged in 2dx MC_Automator [@doi:10.1016/j.jsb.2014.03.016] with motioncor [@doi:10.1038/nmeth.2472].
3D reconstruction was done in IMOD [@doi:10.1006/jsbi.1996.0013].
The alignments were done using the automated fiducial tracking function and the 3D reconstructions were done using the weighted back projection followed by a nonlinear anisotropic diffusion (NAD) filtering.
Following tomogram reconstruction only synaptosomes that fulfilled the following criteria were used: 1) even and non-broken PM, 2) synaptic cleft still attached to the presynapse, 3) spherical vesicles, and 4) a mitochondrion in the presynapse necessary to cover the energy demands of the synapse.
These criteria indicate that the synaptosome is functional [@doi:10/b5tzcr].

Cultured mouse neurons tilt series were acquired at a Titan Krios, equipped with a Falcon 3 direct electron detector (Thermofisher Scientific) without energy filter.
The Falcon camera was operated in linear mode. 
Tilt series were acquired using the TEM Tomography software (TFS) for automated acquisition recorded typically from -60° to 60° with a 2° angular increment and an unbinned pixel size of 0.37 nm. 
Defocus was set between -6 to -10 µm and the total electron dose used was about 80-100 e^-^/Å^2^. Tomogram reconstruction was done as for synaptosome datasets. 

### Manual and automatic segmentation procedures
Manual segmentation of SVs, mitochondria, and the active zone PM was done in IMOD (Figure S4A&B).
The boundary marked the region to be analyzed by Pyto [@doi:10.1016/j.jsb.2016.10.004].
 The analysis by Pyto was essentially the same as described previously [@doi:10.1083/jcb.200908082] [@doi:10.1016/j.jsb.2016.10.004].
In short, the segmented area is divided in 1 voxel thick layers parallel to the active zone for distance calculations.
A hierarchical connectivity segmentation detects densities interconnecting vesicles (so called connectors) and densities connecting vesicles to the active zone PM (so called tethers) (Figure S4C).
Distance calculations are done with the center of the vesicle.
Mainly default settings were used.
The segmentation procedure is conservative and tends to miss some tethers and connectors because of noise.
Consequently, the numbers of tethers and connectors should not be considered as absolute values but rather to compare experimental groups.
All tomograms analyzed by Pyto were obtained on the same microscope with the same tilt range.
The margin of error for false negatives and positives was found to be less than 10% by comparison with ground truth [@doi:10.1016/j.jsb.2016.10.004].
As it was done before, an upper limit was set between 2100 and 3200 nm3 on segment volume.
The tomograms that were used for this analysis were binned by a factor of 2 to 3, resulting in voxel sizes between 2.1 and 2.4 nm.
Tether and connector length were calculated using the midpoint method [@doi:10.1016/j.jsb.2016.10.004].
From the stimulated synaptosomes only those that showed visible signs of exocytosis where used for analysis in Pyto.

### Data analysis
If not stated otherwise data in the text are described as mean ± standard error to the mean (SEM).
Wherever possible, data were presented as box plots with the following settings: orange bar, median; box extremities, lower and upper quartiles; whiskers extend up to 1.5 x interquartile range; dots, outliers.
We used the same statistical tests as in [@doi:10.1083/jcb.200908082; @doi:10.1083/jcb.201206063].
For normal distributed data, the Student’s t-test was used.
*__remove the following sentence if necessary__* For data deviating from the normal distribution, the Kruskal-Wallis (K-W) test was used.
In addition, for data that required to be split into discrete bins (e.g. fraction of connected vesicles by distance to active zone), the χ2 test was used.
To calculate the correlation coefficient for paired samples (such as vesicle distance to active zone and number of tethers), Spearman’s rank correlation was used (ρ-test).
The confidence values were calculated using two-tailed tests and were indicated in the graphs by *, P<0.05; **, P<0.01; ***, P<0.001.
We did not apply statistical methods to predetermine sample size but similar sample sizes as previously reported have been used [@doi:10.1083/jcb.200908082].
It was not necessary to apply randomization.


## References {.page_break_before}

<!-- Explicitly insert bibliography here -->
<div id="refs"></div>
